charmhelpers/core/host.py

```diff
@@ -730,37 +730,77 @@ def wrapped_f(*args, **kwargs):
 
 
 def restart_on_change_helper(lambda_f, restart_map, stopstart=False,
-                             restart_functions=None):
+                             restart_functions=None,
+                             restart_gate_f=None,
+                             post_svc_restart_f=None,
+                             pre_restarts_wait_f=None):
     """Helper function to perform the restart_on_change function.
 
     This is provided for decorators to restart services if files described
     in the restart_map have changed after an invocation of lambda_f().
 
+    This functions allows for a number of helper functions to be passed.
+
+    `restart_functions` is a map with a service as the key and the
+    corresponding value being the function to call to restart the service. For
+    example if `restart_functions={'some-service': my_restart_func}` then
+    `my_restart_func` should a function which takes one argument which is the
+    service name to be retstarted.
+
+    `restart_gate_f` is a function which checks that a restart is permitted. It
+    should returna bool which indicates if a restart is allowed and should
+    take a service name (str) and a list of changed files (List[str]) as
+    arguments.
+
+    `post_svc_restart_f` is a function which runs after a service has been
+    restarted. It takes the service name that was restarted as an argument.
+
+    `pre_restarts_wait_f` is a function which is called before any restarts
+    occur. The use case for this is an application which wants to try and
+    stagger restarts between units.
+
     @param lambda_f: function to call.
     @param restart_map: {file: [service, ...]}
     @param stopstart: whether to stop, start or restart a service
     @param restart_functions: nonstandard functions to use to restart services
                               {svc: func, ...}
+    @param restart_gate_f: A function used to check if the restart is
+                            permitted.
+    @param post_svc_restart_f: A function run after a service has
+                               restarted.
+    @param pre_restarts_wait_f: A function callled before any restarts.
```

Comment:
I'd switch these all (and the previous ones) to `:param xyz:` and `:type xyz:` so that the functions can be declared.  e.g. `:type restart_gate_f` Callable[[str], boolean]`.  Also the name (I think) would be better as `can_restart_now_f`  .. i.e. it's used as a inverted check (returns False if you can't restart)?